core/src/message.rs

```diff
@@ -13,40 +13,111 @@
+///
+/// [crate::Project::members] is initialized with the transaction author as the only member.
+///
+/// [crate::Project::account_id] is generated randomly.
+///
+/// # State-dependent validations
+///
+/// A project with the same ID must not yet exist.
+///
+/// A checkpoint with the given ID must exist.
 #[derive(Decode, Encode, Clone, Debug, Eq, PartialEq)]
 pub struct RegisterProject {
     pub id: ProjectId,
+
+    /// Initial checkpoint of the project.
     pub checkpoint_id: CheckpointId,
+
+    /// Opaque metadata that cannot be changed.
+    ///
+    /// Used by the application.
     pub metadata: Bytes128,
 }
 
+/// Add a new checkpoint to the state.
+///
+/// # State changes
+///
+/// If successful, adds a new [crate::Checkpoint] with the given parameters to the state.
+///
+/// # State-dependent validations
+///
+/// If `previous_checkpoint_id` is provided, a checkpoint with the given ID must exist in the state.
 #[derive(Decode, Encode, Clone, Debug, Eq, PartialEq)]
 pub struct CreateCheckpoint {
     pub project_hash: H256,
     pub previous_checkpoint_id: Option<CheckpointId>,
 }
 
+/// Updates [crate::Project::current_cp].
+///
+/// # State changes
+///
+/// If successful, adds a new [crate::Checkpoint] with the given parameters to the state.
+///
+/// # State-dependent validations
+///
+/// The project `project_id` must exist.
+///
+/// The checkpoint `new_checkpoint_id` must exist.
+///
+/// The transaction author must be in [crate::Project::members] of the given project.
 #[derive(Decode, Encode, Clone, Debug, Eq, PartialEq)]
 pub struct SetCheckpoint {
     pub project_id: ProjectId,
     pub new_checkpoint_id: CheckpointId,
 }
 
+/// Transfer funds from a project account to an account
+///
+/// # State changes
+///
+/// If successful, `balance` is deducated from the project account and added to the the recipient
+/// account. The project account is given by [crate::Project::account_id] of the given project.
+///
+/// If the recipient account did not exist before, it is created. The recipient account may be a
+/// user account or a project account.
+///
+/// # State-dependent validations
+///
+/// The author must be a member of [crate::Project::members].
+///
+/// The project account must have a balance of at least `balance`.
 #[derive(Decode, Encode, Clone, Debug, Eq, PartialEq)]
 pub struct TransferFromProject {
     pub project: ProjectId,
     pub recipient: AccountId,
     pub value: Balance,
 }
 
+/// Transfer funds from one account to another.
+///
+/// # State changes
+///
+/// If successful, `balance` is deducated from the transaction author account and added to the the
+/// recipient account. If the recipient account did not exist before, it is created.
+///
+/// The recipient account may be a user account or a project account.
+///
+/// # State-dependent validations
+///
+/// The author account must have a balance of at least `balance`.
```

Comment:
Here and above, maybe disambiguate to

```suggestion
/// The author account must have a free balance of at least `balance`.
```

?